rp2040-hal/src/i2c/controller.rs

```diff
@@ -458,3 +481,16 @@ impl<A: ValidAddress, T: Deref<Target = Block>, PINS> eh1::I2c<A> for I2C<T, PIN
         self.transaction(address, operations.iter_mut())
     }
 }
+
+#[cfg(feature = "i2c-write-iter")]
+impl<A: i2c_write_iter::AddressMode + ValidAddress, T: Deref<Target = Block>, PINS>
+    i2c_write_iter::I2cIter<A> for I2C<T, PINS, Controller>
+{
+    fn transaction_iter<'a, O, B>(&mut self, address: A, operations: O) -> Result<(), Self::Error>
+    where
+        O: IntoIterator<Item = i2c_write_iter::Operation<'a, B>>,
+        B: IntoIterator<Item = u8>,
```

Comment:
This signature causes a (IMHO) surprising limitation: All Operations must use exactly the same type B.
So for example, this doesn't work:
```
        controller
            .transaction_iter(
                addr,
                [
                    i2c_write_iter::Operation::WriteIter(
                        [1, 2, 3],
                    ),
                    i2c_write_iter::Operation::WriteIter(
                        [1, 2, 3, 4],
                    ),
                ]
                .into_iter(),
            )
            .await
```

The error message is:
```
error[E0308]: mismatched types
   --> on-target-tests/tests/i2c_tests/non_blocking.rs:379:25
    |
378 |                     i2c_write_iter::Operation::WriteIter(
    |                     ------------------------------------ arguments to this enum variant are incorrect
379 |                         [1, 2, 3, 4],
    |                         ^^^^^^^^^^^^ expected an array with a fixed size of 3 elements, found one with 4 elements
    |
help: the type constructed contains `[u8; 4]` due to the type of the argument passed
   --> on-target-tests/tests/i2c_tests/non_blocking.rs:378:21
    |
378 | /                     i2c_write_iter::Operation::WriteIter(
379 | |                         [1, 2, 3, 4],
    | |                         ------------ this argument influences the type of `i2c_write_iter`
380 | |                         // samples.clone().into_iter(),
381 | |                     ),
    | |_____________________^
```

As you can see I just modified the on-target-test so send two arrays of different sizes.
Am I missing something or does this limit the usefulness of the iterator based API?

Reply:
This should be easily worked around using a slice rather than an array:

```rust
controller
  .transaction_iter(
    addr,
    [
      i2c_write_iter::Operation::WriteIter(
        [1, 2, 3][..].iter().cloned(),
      ),
      i2c_write_iter::Operation::WriteIter(
        [1, 2, 3, 4][..].iter().cloned(),
      ),
    ]
    .into_iter(),
  )
  .await
```

This might not be the smoothest, but it should work.
I’m also open to alternatives :)

Reply:
See: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a1851ce0c724a18df386fe3bb913ae93

Some level of `Either` (or equivalent eg with more variants) can be used too, as long as they are `IntoIterator<Item = u8>`.

Reply:
> This should be easily worked around using a slice rather than an array:

Sorry, while this obviously works, it's not easy. :smiling_face_with_tear: 

I don't know a good alternative. If it was easy, I'm sure someone would have found a solution in https://github.com/rust-embedded/embedded-hal/issues/367 and the function would still be in embedded-hal.